### PR TITLE
Fix `complex` constructor

### PR DESCRIPTION
**Pull request by smlerman**
Fix for issue https://github.com/python/typeshed/issues/8440

### CODE REVIEW

**Comment by Akuli** [comment]
Lol, we both changed the title :)

File: stdlib/builtins.pyi [hunk 1]
```diff
@@ -364,10 +364,24 @@ class float:
...
+            cls: type[Self],
+            real: complex | SupportsComplex | SupportsFloat | SupportsIndex = ...,
+            imag: complex | SupportsFloat | SupportsIndex = ...,
+        ) -> Self: ...
+        @overload
+        def __new__(cls: type[Self], real: str | SupportsComplex | SupportsFloat | SupportsIndex | complex) -> Self: ...
+    else:
+        @overload
+        def __new__(
+            cls: type[Self], real: complex | SupportsComplex | SupportsFloat = ..., imag: complex | SupportsFloat = ...
```
Comment (Akuli) [inline stdlib/builtins.pyi:380]: Also accepts `int`, but that's included implicitly in `complex`.

File: stdlib/builtins.pyi [hunk 2]
```diff
@@ -364,10 +364,24 @@ class float:
...
-    @overload
-    def __new__(cls: type[Self], real: float = ..., imag: float = ...) -> Self: ...
-    @overload
-    def __new__(cls: type[Self], real: str | SupportsComplex | SupportsIndex | complex) -> Self: ...
+    if sys.version_info >= (3, 8):
+        # Python doesn't currently accept SupportsComplex for the second argument
+        @overload
+        def __new__(
+            cls: type[Self],
+            real: complex | SupportsComplex | SupportsFloat | SupportsIndex = ...,
```
Comment (Akuli) [inline stdlib/builtins.pyi:372]: I find it weird that `SupportsComplex` doesn't include `complex`.

**Review by Akuli — APPROVED** [review]
Thanks! I used a script and all Python versions 3.6 to 3.10 to verify this.

<details>
<summary>my script</summary>

```python3
class SupportsFloat:
    def __float__(self):
        return 1.0

class SupportsComplex:
    def __complex__(self):
        return 1+2j

class SupportsIndex:
    def __index__(self):
        return 1

objects = {
    "SupportsFloat": SupportsFloat(),
    "SupportsComplex": SupportsComplex(),
    "SupportsIndex": SupportsIndex(),
    "float": 1.0,
    "complex": 1+2j,
    "int": 1,
}

#single_arg = []
#for name, obj in objects.items():
#    try:
#        complex(obj)
#    except Exception:
#        pass
#    else:
#        single_arg.append(name)
#print(single_arg)

double_arg = []
for name1, obj1 in objects.items():
    for name2, obj2 in objects.items():
        try:
            complex(obj1, obj2)
        except Exception:
            pass
        else:
            double_arg.append((name1, name2))

import pprint
pprint.pprint(double_arg)
```

</details>

File: stdlib/builtins.pyi (hunk 2 above)
Comment (AlexWaygood) [reply stdlib/builtins.pyi:372]: You're not the only one — I believe this specific oddity was why they added `complex.__complex__` in 3.11 :)

**Pull request merged**